File: xim/resource/table/FTable.kt
package xim.resource.table

import xim.resource.ByteReader
import xim.poc.browser.DatLoader
import xim.util.OnceLogger
import kotlin.experimental.or

class VTable(val byteReader: ByteReader) {
    fun getVersion(fileId: Int): Int? {
        if (fileId < 0 || fileId > byteReader.bytes.length) {
            OnceLogger.error("FileId is invalid: ${fileId.toString(0x10)}")
            return null
        }

        return byteReader.bytes[fileId].toInt()
    }
}

class FTable(val byteReader: ByteReader) {
    fun getFile(fileId: Int): Pair<Int, Int> {
        byteReader.position = 2 * fileId
        val fileValue = byteReader.next16()

        val fileName = fileValue and 0x7F
        val folderName = fileValue ushr 7
        return Pair(folderName, fileName)
    }
}

object FileTableManager: LoadableResource {
    private const val numTables = 9

    private lateinit var vTable: VTable
    private var loadedVTables = 0

    private lateinit var fTable: FTable
    private var loadedFTables = 0

    private var preloaded = false

    fun getFilePath(fileId: Int?) : String? {
        if (fileId == null) { return null }

        val versionNumber = vTable.getVersion(fileId) ?: return null
        if (versionNumber == 0) { return null }
        val versionString = if (versionNumber == 1) { "" } else { versionNumber.toString() }

        val (folderNumber, fileNumber) = fTable.getFile(fileId)

        return "ROM$versionString/$folderNumber/$fileNumber.DAT"
    }

    override fun preload() {
        if (preloaded) { return }
        preloaded = true

        for (i in 1 .. numTables) {
            loadTable(i)
        }
    }

    override fun isFullyLoaded() : Boolean {
        return loadedVTables == numTables && loadedFTables == numTables
    }

    private fun loadTable(tableIndex: Int) {
        val prefix = if (tableIndex == 1) { "" } else { "ROM${tableIndex}/" }
        val postfix = if (tableIndex == 1) { "" } else { tableIndex.toString() }
        DatLoader.load("${prefix}VTABLE${postfix}.DAT").onReady { onVTableLoad(it.getAsBytes()) }
        DatLoader.load("${prefix}FTABLE${postfix}.DAT").onReady { onFTableLoad(it.getAsBytes()) }
    }

    private fun onVTableLoad(byteReader: ByteReader) {
        if (!this::vTable.isInitialized) { vTable = VTable(byteReader) } else { combine(byteReader, vTable.byteReader) }
        loadedVTables += 1
    }

    private fun onFTableLoad(byteReader: ByteReader) {
        if (!this::fTable.isInitialized) { fTable = FTable(byteReader) } else { combine(byteReader, fTable.byteReader) }
        loadedFTables += 1
    }

    private fun combine(source: ByteReader, destination: ByteReader) {
        for (i in 0 until source.bytes.length) {
            destination.bytes[i] = destination.bytes[i] or source.bytes[i]
        }
    }

}

File: lib/resource/table/fTable.ts
import { ByteReader } from '../byteReader'
import type { LoadableResource } from './loadableResource'

export class VTable {
  readonly byteReader: ByteReader

  constructor(byteReader: ByteReader) {
    this.byteReader = byteReader
  }

  getVersion(fileId: number): number | null {
    if (fileId < 0 || fileId >= this.byteReader.bytes.length) {
      return null
    }

    return this.byteReader.bytes[fileId] ?? null
  }
}

export interface FileTuple {
  readonly folderName: number
  readonly fileName: number
}

export class FTable {
  readonly byteReader: ByteReader

  constructor(byteReader: ByteReader) {
    this.byteReader = byteReader
  }

  getFile(fileId: number): FileTuple {
    this.byteReader.position = 2 * fileId
    const fileValue = this.byteReader.next16()

    return {
      fileName: fileValue & 0x7f,
      folderName: fileValue >>> 7,
    }
  }
}

export class FileTableManager implements LoadableResource {
  private readonly numTables: number
  private readonly loadFile: (path: string) => Promise<Uint8Array>

  private vTable: VTable | null = null
  private loadedVTables = 0

  private fTable: FTable | null = null
  private loadedFTables = 0

  private preloaded = false

  constructor(loadFile: (path: string) => Promise<Uint8Array>, numTables = 9) {
    this.loadFile = loadFile
    this.numTables = numTables
  }

  async preload(): Promise<void> {
    if (this.preloaded) {
      return
    }
    this.preloaded = true

    await Promise.all(Array.from({ length: this.numTables }, (_, i) => this.loadTable(i + 1)))
  }

  isFullyLoaded(): boolean {
    return this.loadedVTables === this.numTables && this.loadedFTables === this.numTables
  }

  getFilePath(fileId: number | null): string | null {
    if (fileId === null || this.vTable === null || this.fTable === null) {
      return null
    }

    const versionNumber = this.vTable.getVersion(fileId)
    if (versionNumber === null || versionNumber === 0) {
      return null
    }

    const versionString = versionNumber === 1 ? '' : String(versionNumber)
    const { folderName, fileName } = this.fTable.getFile(fileId)
    return `ROM${versionString}/${folderName}/${fileName}.DAT`
  }

  private async loadTable(tableIndex: number): Promise<void> {
    const prefix = tableIndex === 1 ? '' : `ROM${tableIndex}/`
    const postfix = tableIndex === 1 ? '' : String(tableIndex)

    const [vTableBytes, fTableBytes] = await Promise.all([
      this.loadFile(`${prefix}VTABLE${postfix}.DAT`),
      this.loadFile(`${prefix}FTABLE${postfix}.DAT`),
    ])

    this.onVTableLoad(new ByteReader(vTableBytes))
    this.onFTableLoad(new ByteReader(fTableBytes))
  }

  private onVTableLoad(byteReader: ByteReader): void {
    if (this.vTable === null) {
      this.vTable = new VTable(byteReader)
    } else {
      this.combine(byteReader, this.vTable.byteReader)
    }

    this.loadedVTables += 1
  }

  private onFTableLoad(byteReader: ByteReader): void {
    if (this.fTable === null) {
      this.fTable = new FTable(byteReader)
    } else {
      this.combine(byteReader, this.fTable.byteReader)
    }

    this.loadedFTables += 1
  }

  private combine(source: ByteReader, destination: ByteReader): void {
    const max = Math.min(source.bytes.length, destination.bytes.length)
    for (let i = 0; i < max; i += 1) {
      destination.bytes[i] = (destination.bytes[i] ?? 0) | (source.bytes[i] ?? 0)
    }
  }
}
